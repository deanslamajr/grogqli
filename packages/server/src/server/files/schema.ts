import shortid from 'shortid';
import editJsonFile from 'edit-json-file';
import { IntrospectionQuery } from 'graphql';
import {
  CreateTemporaryOperationRecording,
  SchemasMappingsInput,
} from '@grogqli/schema';
import fs from 'fs';

import {
  doesFileExist,
  getTemporarySchemaRecordingFilename,
  mapObjectToJsonFile,
  SCHEMA_FILE_VERSION,
  getSchemaRecordingFilePath,
  TEMP_SCHEMA_FILE_VERSION,
} from './';

type TemporarySchemaRecording = TemporarySchemaRecordingVersion1;
interface TemporarySchemaRecordingVersion1 {
  version: 1;
  hash: string;
  introspectionQuery: IntrospectionQuery;
}

type PersistTempSchemaRecording = (
  schemaIntrospectionResult: IntrospectionQuery
) => Promise<string>;

export const persistTempSchemaRecording: PersistTempSchemaRecording = async (
  schemaIntrospectionResult
) => {
  // TODO generate schema hash
  const schemaHash = shortid.generate();

  const newTempSchemaRecording: TemporarySchemaRecording = {
    version: TEMP_SCHEMA_FILE_VERSION,
    hash: schemaHash,
    introspectionQuery: schemaIntrospectionResult,
  };

  const schemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  const newTempSchemaRecordingFile = editJsonFile(schemaRecordingsPath);

  if (!doesFileExist(newTempSchemaRecordingFile)) {
    mapObjectToJsonFile(newTempSchemaRecording, newTempSchemaRecordingFile);
    newTempSchemaRecordingFile.save();
  } else {
    console.log(`Schema with hash:${schemaHash} has already been recorded!`);
  }

  return schemaHash;
};

export type SchemaRecording = SchemaRecordingVersion1;
interface SchemaRecordingVersion1 {
  version: 1;
  id: string;
  name: string;
  hash: string;
  urls: string[];
  introspectionQuery: IntrospectionQuery;
}

export const getSchemaRecordingFile = async (
  schemaId: string
): Promise<SchemaRecording> => {
  const schemaRecordingFilePath = await getSchemaRecordingFilePath(schemaId);

  return JSON.parse(
    await fs.promises.readFile(schemaRecordingFilePath, 'utf8')
  );
};

type CreateNewSchemaRecording = (params: {
  schemaUrl: string;
  schemaHash: string;
}) => Promise<string>;
const createNewSchemaRecording: CreateNewSchemaRecording = async ({
  schemaHash,
  schemaUrl,
}) => {
  const tempSchemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  let temporarySchemaRecordingFile: TemporarySchemaRecording;
  try {
    temporarySchemaRecordingFile = JSON.parse(
      await fs.promises.readFile(tempSchemaRecordingsPath, 'utf8')
    );
  } catch (error) {
    console.error(error);
    throw new Error(`
      Error while creating new schema recording from temporary schema recording:
      Cannot find a file for the given temporary schema recording hash:${schemaHash}
    `);
  }

  const generateUniqueSchemaIdAndCreateSchemaRecordingFile = async (): Promise<SchemaRecording> => {
    const newSchemaRecording: SchemaRecording = {
      version: SCHEMA_FILE_VERSION,
      id: shortid.generate(),
      name: 'tester schema', // TODO replace with user input
      hash: schemaHash,
      urls: [schemaUrl],
      introspectionQuery: temporarySchemaRecordingFile.introspectionQuery,
    };

    const schemaRecordingFilePath = await getSchemaRecordingFilePath(
      newSchemaRecording.id
    );

    const newSchemaRecordingFile = editJsonFile(schemaRecordingFilePath);

    if (doesFileExist(newSchemaRecordingFile)) {
      return generateUniqueSchemaIdAndCreateSchemaRecordingFile();
    } else {
      mapObjectToJsonFile(newSchemaRecording, newSchemaRecordingFile);
      newSchemaRecordingFile.save();
      return newSchemaRecording;
    }
  };

  const newSchemaRecording = await generateUniqueSchemaIdAndCreateSchemaRecordingFile();

  return newSchemaRecording.id;
};

export interface UpdatedSchemasMapping {
  schemaHash: string;
  schemaId: string;
}

type ConditionallyCreateOrUpdateSchemaRecordings = (
  schemaMappings: SchemasMappingsInput[]
) => Promise<UpdatedSchemasMapping[]>;
export const conditionallyCreateOrUpdateSchemaRecordings: ConditionallyCreateOrUpdateSchemaRecordings = async (
  schemaMappings
) => {
  return Promise.all(
    schemaMappings.map<Promise<UpdatedSchemasMapping>>(
      async ({
        opsRecordingsSchemaHash,
        targetSchemaId,
        opsRecordingsSchemaUrl,
      }) => {
        // handle NEW case
        if (targetSchemaId === 'NEW') {
          const newSchemaId = await createNewSchemaRecording({
            schemaUrl: opsRecordingsSchemaUrl,
            schemaHash: opsRecordingsSchemaHash,
          });
          return {
            schemaHash: opsRecordingsSchemaHash,
            schemaId: newSchemaId,
          };
        } else {
          // TODO
          // handle conditionally adding any new opsRecordingsSchemaUrl's to the mapped targetSchemaId
          return {
            schemaHash: opsRecordingsSchemaHash,
            schemaId: targetSchemaId,
          };
        }
      }
    )
  );
};
