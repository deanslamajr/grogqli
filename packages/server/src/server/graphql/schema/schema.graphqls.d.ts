/* 19a699f8a8761279db5b06650f17155319410b52
 * This file is automatically generated by graphql-let. */

import { GraphQLResolveInfo } from 'graphql';
export declare type Maybe<T> = T | null;
export declare type Exact<T extends {
    [key: string]: unknown;
}> = {
    [K in keyof T]: T[K];
};
/** All built-in and custom scalars, mapped to their actual values */
export declare type Scalars = {
    ID: string;
    String: string;
    Boolean: boolean;
    Int: number;
    Float: number;
};
export declare type Recording = {
    __typename?: 'Recording';
    replaceThisBullshxt: Scalars['String'];
};
export declare type Query = {
    __typename?: 'Query';
    getRecordings: Array<Maybe<Recording>>;
};
export declare type ResolverTypeWrapper<T> = Promise<T> | T;
export declare type LegacyStitchingResolver<TResult, TParent, TContext, TArgs> = {
    fragment: string;
    resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};
export declare type NewStitchingResolver<TResult, TParent, TContext, TArgs> = {
    selectionSet: string;
    resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};
export declare type StitchingResolver<TResult, TParent, TContext, TArgs> = LegacyStitchingResolver<TResult, TParent, TContext, TArgs> | NewStitchingResolver<TResult, TParent, TContext, TArgs>;
export declare type Resolver<TResult, TParent = {}, TContext = {}, TArgs = {}> = ResolverFn<TResult, TParent, TContext, TArgs> | StitchingResolver<TResult, TParent, TContext, TArgs>;
export declare type ResolverFn<TResult, TParent, TContext, TArgs> = (parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => Promise<TResult> | TResult;
export declare type SubscriptionSubscribeFn<TResult, TParent, TContext, TArgs> = (parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => AsyncIterator<TResult> | Promise<AsyncIterator<TResult>>;
export declare type SubscriptionResolveFn<TResult, TParent, TContext, TArgs> = (parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => TResult | Promise<TResult>;
export interface SubscriptionSubscriberObject<TResult, TKey extends string, TParent, TContext, TArgs> {
    subscribe: SubscriptionSubscribeFn<{
        [key in TKey]: TResult;
    }, TParent, TContext, TArgs>;
    resolve?: SubscriptionResolveFn<TResult, {
        [key in TKey]: TResult;
    }, TContext, TArgs>;
}
export interface SubscriptionResolverObject<TResult, TParent, TContext, TArgs> {
    subscribe: SubscriptionSubscribeFn<any, TParent, TContext, TArgs>;
    resolve: SubscriptionResolveFn<TResult, any, TContext, TArgs>;
}
export declare type SubscriptionObject<TResult, TKey extends string, TParent, TContext, TArgs> = SubscriptionSubscriberObject<TResult, TKey, TParent, TContext, TArgs> | SubscriptionResolverObject<TResult, TParent, TContext, TArgs>;
export declare type SubscriptionResolver<TResult, TKey extends string, TParent = {}, TContext = {}, TArgs = {}> = ((...args: any[]) => SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>) | SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>;
export declare type TypeResolveFn<TTypes, TParent = {}, TContext = {}> = (parent: TParent, context: TContext, info: GraphQLResolveInfo) => Maybe<TTypes> | Promise<Maybe<TTypes>>;
export declare type IsTypeOfResolverFn<T = {}> = (obj: T, info: GraphQLResolveInfo) => boolean | Promise<boolean>;
export declare type NextResolverFn<T> = () => Promise<T>;
export declare type DirectiveResolverFn<TResult = {}, TParent = {}, TContext = {}, TArgs = {}> = (next: NextResolverFn<TResult>, parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => TResult | Promise<TResult>;
/** Mapping between all available schema types and the resolvers types */
export declare type ResolversTypes = {
    Recording: ResolverTypeWrapper<Recording>;
    String: ResolverTypeWrapper<Scalars['String']>;
    Query: ResolverTypeWrapper<{}>;
    Boolean: ResolverTypeWrapper<Scalars['Boolean']>;
};
/** Mapping between all available schema types and the resolvers parents */
export declare type ResolversParentTypes = {
    Recording: Recording;
    String: Scalars['String'];
    Query: {};
    Boolean: Scalars['Boolean'];
};
export declare type RecordingResolvers<ContextType = any, ParentType extends ResolversParentTypes['Recording'] = ResolversParentTypes['Recording']> = {
    replaceThisBullshxt?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
    __isTypeOf?: IsTypeOfResolverFn<ParentType>;
};
export declare type QueryResolvers<ContextType = any, ParentType extends ResolversParentTypes['Query'] = ResolversParentTypes['Query']> = {
    getRecordings?: Resolver<Array<Maybe<ResolversTypes['Recording']>>, ParentType, ContextType>;
};
export declare type Resolvers<ContextType = any> = {
    Recording?: RecordingResolvers<ContextType>;
    Query?: QueryResolvers<ContextType>;
};
/**
 * @deprecated
 * Use "Resolvers" root object instead. If you wish to get "IResolvers", add "typesPrefix: I" to your config.
 */
export declare type IResolvers<ContextType = any> = Resolvers<ContextType>;

          
// This is an extra code in addition to what graphql-codegen makes.
// Users are likely to use 'graphql-tag/loader' with 'graphql-tag/schema/loader'
// in webpack. This code enables the result to be typed.
import { DocumentNode } from 'graphql'
export default typeof DocumentNode
