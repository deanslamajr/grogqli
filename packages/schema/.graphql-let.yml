schema: 'src/schema/**/*.graphql'
documents: 'src/queries/**/*.graphql'
plugins:
  - typescript
  - typescript-operations
  - typed-document-node
  - typescript-resolvers
schemaEntrypoint: 'src/schema/schema.graphql'
respectGitIgnore: true
cacheDir: src/generated
config:
  avoidOptionals: true
  useIndexSignature: true
