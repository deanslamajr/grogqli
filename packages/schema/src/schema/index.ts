import { typeDefs as sharedTypes } from './shared';
import { typeDefs as queryTypes } from './queries';
import { typeDefs as mutationTypes } from './mutations';

export const typeDefs = [
  ...queryTypes,
  ...mutationTypes,
  ...sharedTypes,
];