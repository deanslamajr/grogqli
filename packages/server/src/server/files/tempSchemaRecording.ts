import {
  buildSchema,
  IntrospectionQuery,
  lexicographicSortSchema,
  GraphQLSchema,
  printSchema,
  stripIgnoredCharacters,
} from 'graphql';
import editJsonFile from 'edit-json-file';
import crypto from 'crypto';
import fs from 'fs';
import sanitize from 'sanitize-filename';

import { buildGraphQLSchemaFromIntrospectionQuery } from '../utils/buildGraphQLSchemaFromIntrospectionQuery';
import { createSchemaSDL } from '../utils/createSchemaSDL';

import {
  doesFileExist,
  getTemporarySchemaRecordingFilename,
  mapObjectToJsonFile,
  TEMP_SCHEMA_FILE_VERSION,
} from './';

export type TemporarySchemaRecording = TemporarySchemaRecordingVersion1;
interface TemporarySchemaRecordingVersion1 {
  version: 1;
  hash: string;
  introspectionQuery: IntrospectionQuery;
}

type Get = (schemaHash) => Promise<TemporarySchemaRecording | null>;
export const get: Get = async (schemaHash) => {
  const tempSchemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  let temporarySchemaRecordingFile: TemporarySchemaRecording | null;
  try {
    temporarySchemaRecordingFile = JSON.parse(
      await fs.promises.readFile(tempSchemaRecordingsPath, 'utf8')
    );
  } catch (error) {
    console.error(
      `could not open temporarySchemaRecordingFile with schemaHash:${schemaHash}`,
      { error }
    );
    temporarySchemaRecordingFile = null;
  }

  return temporarySchemaRecordingFile;
};

type BuildGraphQLSchema = (schemaHash: string) => Promise<GraphQLSchema>;
export const buildGraphQLSchema: BuildGraphQLSchema = async (schemaHash) => {
  const temporarySchemaRecordingFile = await get(schemaHash);
  if (temporarySchemaRecordingFile === null) {
    throw new Error(`
      Error while building a graphql schema:
      Cannot find a file for the given temporary schema recording hash:${schemaHash}
    `);
  }

  return buildGraphQLSchemaFromIntrospectionQuery(
    temporarySchemaRecordingFile.introspectionQuery
  );
};

type PersistTempSchemaRecording = (
  schemaIntrospectionResult: IntrospectionQuery
) => Promise<string>;

export const persistTempSchemaRecording: PersistTempSchemaRecording = async (
  schemaIntrospectionResult
) => {
  const schemaHash = await generateHashFromSchema(schemaIntrospectionResult);

  const newTempSchemaRecording: TemporarySchemaRecording = {
    version: TEMP_SCHEMA_FILE_VERSION,
    hash: schemaHash,
    introspectionQuery: schemaIntrospectionResult,
  };

  const schemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  const newTempSchemaRecordingFile = editJsonFile(schemaRecordingsPath);

  if (!doesFileExist(newTempSchemaRecordingFile)) {
    mapObjectToJsonFile(newTempSchemaRecording, newTempSchemaRecordingFile);
    newTempSchemaRecordingFile.save();
  }

  return schemaHash;
};

// from https://github.com/graphql/graphql-js/blob/998bea680d6e11e1c055a400a887a9539de08f75/src/utilities/__tests__/lexicographicSortSchema-test.js#L10
const sortSDL = (sdl: string): string => {
  const schema = buildSchema(sdl);
  return printSchema(lexicographicSortSchema(schema));
};

const generateHash = (stringToHash: string): string => {
  const rawHash = crypto
    .createHash('sha1')
    .update(stringToHash)
    .digest('base64');
  // need to remove reserved filesystem path characters
  // in order to make the hash safe for use as file/folder names
  return sanitize(rawHash);
};

export const generateHashFromSchema = async (
  schemaIntrospectionResult: IntrospectionQuery
): Promise<string> => {
  // 1st - Create SDL string from introspection result object
  const schemaSdl: string = await createSchemaSDL(schemaIntrospectionResult);
  // 2nd - Sort the schema SDL
  const sortedSchemaSdl: string = sortSDL(schemaSdl);
  // 3rd - Remove any characters from SDL string that do not affect the behavior of the schema
  const cleanedSchemaSdl: string = stripIgnoredCharacters(sortedSchemaSdl);
  // Final - Generate hash
  return generateHash(cleanedSchemaSdl);
};
