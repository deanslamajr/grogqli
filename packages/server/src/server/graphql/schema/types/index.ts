import { join } from 'path'
import { loadFilesSync } from '@graphql-tools/load-files'
import { mergeTypeDefs } from '@graphql-tools/merge'

const loadedFiles = loadFilesSync(join(process.cwd(), '**/*.graphqls'))
export const typeDefs = mergeTypeDefs(loadedFiles)