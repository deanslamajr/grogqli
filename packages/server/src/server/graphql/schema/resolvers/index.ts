import {Resolvers} from '@grogqli/schema';

import { queryResolver } from './queries';
import {mutationResolver } from './mutations';
import {subscriptionResolver} from './subscriptions';

export const resolvers: Partial<Resolvers> = {
  Query: queryResolver,
  Mutation: mutationResolver,
  Subscription: subscriptionResolver
};
