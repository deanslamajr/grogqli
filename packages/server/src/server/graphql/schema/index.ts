import { makeExecutableSchema } from '@graphql-tools/schema';
import { typeDefs } from '@grogqli/schema';

import { resolvers } from './resolvers';

const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

export default schema;
