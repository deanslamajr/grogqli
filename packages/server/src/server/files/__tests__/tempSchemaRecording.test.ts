import { introspectionQuery } from './grogqli/schemas/someSchemaId/schema.json';

import { generateHashFromSchema } from '../tempSchemaRecording';

describe('tempSchemaRecording', () => {
  describe('generateHashFromSchema', () => {
    it('should generate a stable hash', async () => {
      const firstHash = await generateHashFromSchema(introspectionQuery);
      const secondHash = await generateHashFromSchema(introspectionQuery);

      expect(firstHash).toEqual(secondHash);
      expect(firstHash).toMatchSnapshot();
    });
  });
});
