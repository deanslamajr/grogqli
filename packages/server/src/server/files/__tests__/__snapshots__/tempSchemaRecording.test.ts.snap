// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tempSchemaRecording generateHashFromSchema should generate a stable hash 1`] = `"OdKUL9Gjd6B+RDrmLyh+25VjKIE="`;
